cosmoz-sse.js

```diff
@@ -1,230 +1,87 @@
-// @license Copyright (C) 2019 Neovici AB - Apache 2 License
-import { PolymerElement } from '@polymer/polymer/polymer-element';
-
-const SSEHandlers = {
-	text: data => data,
-	json: data => JSON.parse(data)
-};
-
-/**
-	`<cosmoz-sse>` is an element that allows subscribing to server-sent events.
-
-	@customElement
-	@polymer
-	@demo demo/index.html
-*/
-class CosmozSSE extends PolymerElement {
-	static get is() {
-		return 'cosmoz-sse';
-	}
-
-	static get properties() {
-		return {
-			_configuration: {
-				type: Object,
-				computed: '_computeConfiguration(withCredentials)'
-			},
-			/**
-			 * The SSE endpoint.
-			 */
-			url: {
-				type: String
-			},
-
-			/**
-			 * The named events you wish to subscribe to.
-			 */
-			events: {
-				type: Array,
-				value: [],
-				observer: '_onEventsChange'
+// @license Copyright (C) 2020 Neovici AB - Apache 2 License
+import {
+	component, useEffect, useMemo, useState
+} from 'haunted';
+
+const getJson = input => {
+		try {
+			return JSON.parse(input);
+		} catch (e) {
+			// eslint-disable-next-line no-console
+			console.error('cant parse JSON:', input);
+		}
+	},
+	SSEHandlers = {
+		text: data => data,
+		json: data => getJson(data)
+	},
+
+	// eslint-disable-next-line max-lines-per-function
+	CosmozSSE = function ({
+		events, withCredentials = false, handleAs = 'json', url
+	}) {
+		const fireEvent = (type, eventInit) => {
```

Comment:
```suggestion
		const fireEvent = useCallback((type, eventInit) => {
```

Use `useCallback` to memoize the function, otherwise a new function is created on every update.

Reply:
as I understood the docs, we should basically only memoize when we want the function to be comparable/dependency of other things. I guess useCallback() will be called every update otherwise too, which might be more expensive than re-declaring?
Maybe a function in the outside scope that takes a context rather?

Reply:
Calling useCallback every update is very fast, it's basically just a lookup.

You could simplify to:
```js
const fire = useCallback((event) => this.dispachEvent(event), []) // empty dependencies array so it only gets assigned once
```

and do:
```js
fire(new CustomEvent(type, {details: 'x'}));
```

This gives you more flexibility, allocates only once, and binds the correct `this` for good.

Reply:
upon further consideration, you could drop the useCallback and just go with
```
fire = (event) => this.spinachEvent(event)
```
because it's not used as a dependency for other hooks